Shipping Zone Integration - Refactor Shipping Rates API

nima-karimi opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Implements part of #1117.

This PR modifies the following APIs so that they can accept and return multiple shipping rates/methods per country (instead of just returning a single rate):

- `GET/POST mc/shipping/rates` 
- `POST mc/shipping/rates/batch`
- `GET mc/shipping/rates/suggestions`

With this change, the APIs now accept the following data schema:
```JSON
{
  "country_code": "US",
  "rates": [
    {
      "method": "flat_rate",
      "currency": "USD",
      "rate": 50,
      "options": {
        "free_shipping_threshold": "100"
      }
    }
  ]
}
```

The `method` parameter specifies the shipping method and for now, can only be set to `flat_rate`. However, once support for the pickup method is added, it can also be set to `local_pickup`.

The `options` parameter accepts an object that contains the `free_shipping_threshold` parameter. This option can be set for the `flat_rate` shipping method to specify the minimum order value for which the free shipping option should be offered.

To offer an unconditional free shipping method, the `rate` parameter can be set to `0`. The `free_shipping_threshold` option will NOT be considered if the shipping rate parameter is `0`. In other words, we can't set a condition if the shipping is already free for the country.

The `ShippingRateTable` now includes two new columns to store the shipping `method` and its `options` alongside each shipping rate. In order to make this change, I had to modify some of the existing code to be able to handle cases where multiple rows were returned for one country. As an example, the `Query` class now also accepts a `groupby` clause that allows creating grouped queries. Moreover, a new DB migration class is introduced that will update the DB once the plugin is updated.

I've also modified the country code validation method that was used in the shipping rates API and a few other places so that it can differentiate between a single value and an array. A new `get_country_codes_array_validate_callback` method is added that only works on arrays of country codes. This is unrelated to this feature but it was a needed improvement because I noticed that the API accepts array values where it should only accept one singular value.

Note that as part of the changes here, `offers_free_shipping` and `free_shipping_threshold` have been removed from the `/mc/settings` endpoint.

### Detailed test instructions:

**Test the DB migration:**

1. Update the `WC_GLA_VERSION` constant to a higher version than the current one.
2. Go to the `Migration20211228T1640692399::get_applicable_version` method and modify the `x.x.x` version to be equal to or greater than the one you set above.
3. Open any page in your website admin
4. Confirm that the `wp_gla_shipping_rates` table is updated and includes the `options` and `method` columns. The `method` column must be set by default to `flat_rate`


**Test the APIs:**

1. Define multiple shipping zones each containing various regions and shipping methods. Make sure you have at least one free shipping method with a minimum order amount option.
2. Make a `GET` request to the `mc/shipping/rates/suggestions` API with a `country_code` parameter and confirm the rate that is returned is the same that you defined in your shipping zone settings.
3. Make a `POST` request to the `mc/shipping/rates` containing the following JSON data:

```JSON
{
  "country_code": "DE",
  "rates": [
    {
      "method": "flat_rate",
      "currency": "USD",
      "rate": 50,
      "options": {
        "free_shipping_threshold": "100"
      }
    }
  ]
}
```

4. Make two `GET` requests to the `mc/shipping/rates/US` and `mc/shipping/rates` APIs and confirm the shipping rate you set above is returned correctly by these APIs.
5. Make a `POST` request to the `mc/shipping/rates/batch` API and send the following JSON data:

```JSON
{
  "country_codes": [
    "DE",
    "FR"
  ],
  "rates": [
    {
      "method": "flat_rate",
      "currency": "USD",
      "rate": 30,
      "options": {
        "free_shipping_threshold": "150"
      }
    }
  ]
}
```

6. Repeat step 4
7. Make a `POST` request to the `wc/gla/mc/settings/sync` API
8. Confirm that the shipping rates you have just set are reflected in your Merchant Center account (https://merchants.google.com/mc/shipping/services). Check the minimum order value as well.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Add - Integration with WooCommerce Shipping Zone to automatically sync shipping settings to Merchant Center

mikkamp commented on src/Internal/DependencyManagement/DBServiceProvider.php:
```diff
@@ -90,6 +91,7 @@ public function register() {
 
 		// Share DB migrations
 		$this->share_migration( MigrationVersion141::class, MerchantIssueTable::class );
+		$this->share_migration( Migration20211228T1640692399::class, ShippingRateTable::class );
```
Is this naming temporary until we have the know the version this will be implemented?

nima-karimi replied:
Not really. I previously used the version number in the class name but that isn't very scalable. We usually don't know the version that this migration is going to be released on and also, for example, if two people are going to be committing a migration class for the same version we would get a conflict. I decided to use a datetime+unix-timestamp naming for them to resolve this.

mikkamp replied:
Ok that makes sense. Although if we are going to have multiple migration classes for a version it would probably be better to merge them.

mikkamp commented on src/DB/Migration/Migration20211228T1640692399.php:
```diff
@@ -0,0 +1,59 @@
+	 * @return string A version number. For example: 1.4.1
+	 */
+	public function get_applicable_version(): string {
+		return 'x.x.x';
```
I assume this is also temporary and will be manually changed, since the version replace in WR won't catch this.

nima-karimi replied:
Hmm. I think we can add a step in the release document to modify this value because there's no way to know this before release.

mikkamp replied:
Yes that sounds like a good idea.

nima-karimi replied:
Added a new step to the preparing for release section of our wiki: https://github.com/woocommerce/google-listings-and-ads/wiki/Release-Process

mikkamp commented on src/DB/Query.php:
```diff
@@ -287,8 +306,14 @@ protected function build_query( bool $get_count = false ): string {
 		if ( ! $get_count ) {
-			$pieces[] = "GROUP BY `{$this->table->get_name()}`.`{$this->table->get_primary_column()}`";
+			if ( empty( $this->groupby ) ) {
+				$pieces[] = "GROUP BY `{$this->table->get_name()}`.`{$this->table->get_primary_column()}`";
```
I know we discussed this before somewhere (I think when the get_count functionality was added), but it seems it was still left in the code.

All the primary columns in our tables are unique, so grouping by a unique column doesn't change the result. Do we still want to leave this default group by in place?

nima-karimi replied:
I also don't think that's necessary but just for the sake of not making this PR more complicated, I left it as is. I'm not sure if it'll have any side effects but it's defined in a common class so maybe it's best to change that in another PR so that we can test it appropriately.

mikkamp commented on src/API/Google/Settings.php:
```diff
@@ -244,15 +228,16 @@ protected function get_options_object(): OptionsInterface {
 	 * @return Service
 	 */
-	protected function create_main_service( string $country, string $currency, $rate ): Service {
+	protected function create_shipping_service( string $country, string $method, string $currency, $rate ): Service {
```
I guess this was already present in the code, but I was just wondering why we have `$rate` as mixed. In the DB it's explicitly marked as a double. Should we be a bit stricter here.

nima-karimi replied:
Yes, that confused me as well. I'm going to mark it as a `float` just to be in line with other places.

nima-karimi replied:
Done in https://github.com/woocommerce/google-listings-and-ads/pull/1171/commits/0e9904992fad55b5763b6fa1f65f47ce4783c7b1

mikkamp commented on src/DB/Query/ShippingRateQuery.php:
```diff
@@ -41,6 +41,28 @@ protected function sanitize_value( string $column, $value ) {
 		}
 
+		if ( 'options' === $column ) {
+			$value = maybe_serialize( $value );
```
This isn't fully sanitizing as the name of the function implies. Any data could be inserted once serialized, not just a set of options. Do we need to check here if it's in the format we expect?

nima-karimi replied:
I don't think we need to check the exact format because that gets a bit outside the responsibility of this class. I originally wanted to define "types" for each of the columns so that we can properly validate and sanitize them but that would make these query/table classes more complicated and push them towards being an ORM.

I added a check (https://github.com/woocommerce/google-listings-and-ads/pull/1171/commits/5ee6cdb3ae57a239e0a4bef460fc13e7f6cb8ace) to make sure that the provided value is an array and throw an exception otherwise. However, this still sounds like a "validation" and not "sanitization". Even though the line above does a similar check for the `id` column.

mikkamp replied:
Yes I agree, none of the classes are really doing sanitization in that function. Renaming it might be clearer but at this state it's probably better to leave it as is.

mikkamp reviewed (commented):
At initial glance these changes look good. I ran out of time to complete the full testing of all the changes (I'll pick that up again later). I just wanted to add some comments of things I noticed while going through the code.

One thing that's still not quite clear to me is how it works when they have multiple rates set per country (ex. one per state). Are we still going to save just one flat rate per country? Since a shipping service now supports multiple rates will they all get saved? I seem to be missing code that prevents this from happening or where it's tested in the unit tests.

nima-karimi commented:
> how it works when they have multiple rates set per country (ex. one per state). Are we still going to save just one flat rate per country? Since a shipping service now supports multiple rates will they all get saved? I seem to be missing code that prevents this from happening or where it's tested in the unit tests.

We expect one rate per each `country` and `method`. And in this PR the `pickup` method is not supported so that means we only have one method `flat-rate`. Therefore, each country still has only one shipping method associated with it.

There's a check for existing values in `ShippingRateController::get_create_rate_callback` that prevents storing duplicates in that table.

> ex. one per state

In this version, we are only working on a country level so the user cannot set multiple rates for the same method for one country. We probably need to modify the shipping rate table once more when we want to store multiple rates for states, postal codes, etc. 

Now that I think of it, it might be a good idea to implement those changes to DB with this release so that we don't have to make other changes to the DB in the next version. What do you think?

nima-karimi commented:
This is ready for another review, @mikkamp 🙏

mikkamp commented:
> Now that I think of it, it might be a good idea to implement those changes to DB with this release so that we don't have to make other changes to the DB in the next version. What do you think?

I find it quite limiting in the Google API that only one rate per country can be defined (in the case of a European country that's quite common, but it's not the same to ship across all of the United States with one rate). If we are certain what DB changes are needed to support multiple rates then it would make sense to change them now. Although it sounds like Google would need to change their API to allow this change, so we couldn't really predict what shipping rate structure they would change to.

mikkamp commented:
> There's a check for existing values in ShippingRateController::get_create_rate_callback that prevents storing duplicates in that table.

Thanks for the clarification, it makes more sense now that I look at that code. Although isn't this logic better placed within the `ShippingRateQuery` class? I realize we can't use the `update_or_insert` function because the duplicate is not based on the primary key, however it would feel better placed if the controller just passes a rate to the query class and within the class it determines whether it should be updated or inserted.

mikkamp commented on src/DB/Table/ShippingRateTable.php:
```diff
@@ -26,10 +26,13 @@ protected function get_install_query(): string {
 CREATE TABLE `{$this->get_sql_safe_name()}` (
     id bigint(20) NOT NULL AUTO_INCREMENT,
     country varchar(2) NOT NULL,
+    method varchar(30) NOT NULL DEFAULT 'flat_rate',
```
This looks rather confusing, because this is generally the source of truth on what the DB schema looks like, but in the migration code we drop the default. I understand why it's done, to automatically set a default for the already existing rates. However I wonder if there isn't a clearer method of upgrading this table.

In WC core I see examples of added text fields with `NOT NULL DEFAULT ''`
Would it make more sense to do the same and then set it to a default of `flat_rate` in the migrate class?

nima-karimi replied:
> Would it make more sense to do the same and then set it to a default of flat_rate in the migrate class?

I initially intended to do that but this sounded like a much simpler and fool-proof way so I went with a default. I do agree that updating it manually would make the table definition look better.

But even if we set the default as `''` it would mean that the table definition is wrong because there should never be an empty value set in the `method` field. In this case, the actual default value that we expect is `flat_rate`.

We can skip dropping the `DEFAULT` in the migration class as you suggested but still keep the default as `flat_rate`. Although I'm not sure if that would be the right call because we don't expect a `NOT NULL` column to have a default.

I guess either way we're going to have some minor inconsistency between the Table class and what the real DB looks like. It's just a question of which one is less important.

Another way is to not set a `DEFAULT` at all and still update it manually in the migration class. This will keep the `method` column still as `null` while it's set to be `NOT NULL`. Even though this works on my local MariaDB instance, I'm not sure if that'll work with all versions of MySQL that our users might use.

Let me know what you think.

mikkamp replied:
Thanks for looking into this. I agree I went back and forth quite a few times on possible solutions, but I didn't really find any that fit nicely. I think relying on it becoming NULL when it's set to NOT NULL would be risky (although I haven't tested that much).

I'd say to just run with what you have and add a small note to the table definition that the default will be dropped.

mikkamp reviewed (commented):
Overall the changes look good and they are working as intended. I've tested the API endpoints and they are working well.

I'm not too certain about the way the migration routine is run though. Seems like we could improve or clarify what's happening there a bit more.

nima-karimi commented:
> I find it quite limiting in the Google API that only one rate per country can be defined

It's actually already implemented in Google API but we decided to go at it step by step and release an MVP by the end of 2021 (or in this case, early 2022) because mapping the postal codes and regions would've made this a bit complex.

I think it's still a good idea to do this gradually. We can always add migrations to modify the DB as we go. Just don't want to delay the MVP release any further.

nima-karimi commented:
> Although isn't this logic better placed within the ShippingRateQuery class?

I think it's best to move all this logic into a separate service because the controller is getting huge. That said, this was already in place and I didn't want to modify this part any more than I should. Even though the name of this PR starts with "Refactor" I only wanted to focus on getting the shipping integration working 😅 Refactoring the other parts would be best done in its own PR I think.

nima-karimi commented:
Ready for another round of review, @mikkamp  🙏

mikkamp reviewed (approved):
Thanks for the additional clarifications. I think this is good to go then, we can move any additional refactors to followup PR's.

I'd suggest to just keep an eye on when this will be released and double mention changing the version in the migration class so it doesn't get missed.

ecgan commented:
@nima-karimi , from the way I see it in the PR description, the API changes are breaking changes, and changes on the front-end are needed to be shipped together with the API changes here, right? In that case, I think it would be better to merge this PR into a separate feature branch instead of `develop` branch?

nima-karimi commented:
@ecgan, yes this is a breaking change. I think you can base your PR off of this one and we can wait until the front-end is done before merging this. I need to make some changes to this PR to implement the latest design changes so it'd be good if you also use this as a base branch.

ecgan commented:
@nima-karimi ,

> ... it'd be good if you also use this as a base branch.

Yes, I will make the front-end changes based on your branch anyway. 

My concerns with this PR targeting `develop` branch are:

- I think the front-end is going to take quite some time, which means that when you are done with your changes in this PR, your PR has to remain open until the front-end is done.
- When the front-end PR is done and it gets merged into this PR, your PR here would have both API changes _and_ the front-end changes too, which IMO may not be desired because this PR already has 16 files changed, and the PR title and description are about the API changes.

An idea to address the above concerns is to: 

1. Create a new branch (e.g. `feature/main-shipping-rates-integration`) based on `develop` branch.
2. Change this PR target branch so that it will merge into the new branch.
    - This means that when your PR is done, you can merge it any time, without the need of waiting for the front-end. 
    - The PR will contain API changes only.
3. I will create front-end PR targeting your branch.
    - The PR will contain front-end changes only.
    - Eventually the front-end will be merged into `feature/main-shipping-rates-integration`. 
4. Create a final PR containing both API and front-end changes to merge `feature/main-shipping-rates-integration` into `develop`.
    - This PR should contain other smaller PRs only.

For reference, this is also the approach that I used previously in PR https://github.com/woocommerce/google-listings-and-ads/pull/1119, it is a big PR containing many other smaller PRs, which allows me to make changes in small chunks and ease the PR reviews, and only merge the whole thing into `develop` when everything is done.

What do you think? 

---

> yes this is a breaking change.

I just noticed this: could the PR title be changed to reflect the changes in this PR, because it is not really a "refactor"? 😅

nima-karimi commented:
I understand your concern but I have already merged one PR into this one and also have another pointing into this. I think it's possible to use this one as a base branch because it's already approved and there are no other changes planned for it. What do you think?

nima-karimi commented:
@ecgan oh and the API for submitting and suggesting shipping class rates are removed in https://github.com/woocommerce/google-listings-and-ads/pull/1192 based on the new design changes. Please don't use it in the front-end because once that PR gets merged it will be removed.

ecgan commented:
@nima-karimi ,

> I understand your concern.

Alright then. 😄 

> the API for submitting and _suggesting shipping class rates_ are removed in #1192 based on the new design changes. Please don't use it in the front-end because once that PR gets merged it will be removed.

Sorry, I think I am a bit confused with the "suggesting" and "shipping class" terms. Are you referring to the shipping rates suggestions API (`GET mc/shipping/rates/suggestions`)? I tested the branch in #1192 and the API is returning `success` data for me.

nima-karimi commented:
> Are you referring to the shipping rates suggestions API (GET mc/shipping/rates/suggestions)? I tested the branch in #1192 and the API is returning success data for me.

Oh yeah, I should've been more clear sorry 😅 So the API is still there but it previously returned the rates for each shipping class but now I've removed it. Since we decided to only update the shipping class rates automatically and don't ask the user for input.

ecgan commented:
I noticed that `offers_free_shipping` and `free_shipping_threshold` have been removed from the `/mc/settings` endpoint (see https://github.com/woocommerce/google-listings-and-ads/pull/1171/files#diff-63fb453c2595786bfc37a45d2e3440200cf4898b5839056da6b6fde901e8fab2L103-L118). I have added this info into the PR description above.

ecgan committed:
Refetch shipping rates after upsert.

We need this because some of the shipping rates may be newly created and do not have the id, so we refetch and store them in wp-data.

ecgan committed:
wp-data changes for deleteShippingRates.

ecgan committed:
Change Setup MC AddRateModal and EditRateModal to work with upsertShippingRates and deleteShippingRates.

Commit:
Remove free shipping options from MC settings

These options are now managed in the shipping settings for each country separately, and they do not need to be set globally anymore.

Commit:
Clarify what the `parse_shipping_zones` method does in comment

nima-karimi committed:
Merge pull request #1192 from woocommerce/feature/add-automatic-setting-for-shipping

Add automatic settings for shipping rates

nima-karimi committed:
Merge pull request #1202 from woocommerce/feature/modify-shipping-rate-api

Change the shipping rate API response and request args

ecgan committed:
Merge pull request #1201 from woocommerce/feature/shipping-rates-automatic-settings-ui

Commit:
Merge branch 'feature/refactor-shipping-rates-api' into feature/remove-free-shipping-mc-settings

# Conflicts:
#	src/Tracking/TrackerSnapshot.php
#	src/Value/MerchantCenterSettings.php

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-api-ui-changes

nima-karimi committed:
Merge pull request #1215 from woocommerce/feature/remove-free-shipping-mc-settings

Remove free shipping options from MC settings

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-api-ui-changes

ecgan committed:
Remove offers_free_shipping and free_shipping_threshold settings from front-end.

ecgan committed:
Remove offers_free_shipping and free_shipping_threshold settings from front-end.

(cherry picked from commit 0e91c87131a85d21e76f17b0dfcd22d81476a221)
(cherry picked from commit 9ce176db0b25b9d3811a1eeed8856d2e75cd1136)

ecgan committed:
Change countryCode to country in ShippingCountriesForm.

This is to make the new shippingRate schema work in Edit Free Listings.

ecgan committed:
Refactor ShippingRateSetup.

ShippingRateSetup doesn't need formProps, just need value and onChange, and it does not need to know the field name 'shipping_country_rates'.

ecgan committed:
Move and refactor ShippingRateSetup with FlatShippingRatesInput.

ecgan committed:
Rename FlatShippingRatesInput to FlatShippingRatesInputCards.

ecgan committed:
Remove unneeded EstimatedShippingRatesCard folder.

ecgan committed:
Make upsertShippingRates work for Edit Free Listings.

ecgan committed:
Display green free shipping pill when shipping rate is 0.

ecgan committed:
Make free shipping pill wrap with better layout.

ecgan committed:
Code restructure.

ecgan committed:
Convert shipping rate from string to number.

ecgan committed:
Fix shipping rate options should be object, not array.

ecgan committed:
Display OfferFreeShippingCard.

ecgan committed:
Display MinimumOrderCard.

ecgan committed:
Fix free_shipping_threshold typo.

ecgan committed:
Remove free_shipping_threshold when user selects 'no' for free shipping.

ecgan committed:
Display MinimumOrderCard with MinimumOrderInput.

ecgan committed:
Consolidate 'yes' and 'no' values within OfferFreeShippingCard.

ecgan committed:
Code refactor in MinimumOrderCard.

ecgan committed:
Label and layout for MinimumOrderInput.

ecgan committed:
Code refactor with MinimumOrderInputLabel.

ecgan committed:
Add edit minimum order button and modal.

ecgan committed:
Rename MinimumOrderInput to MinimumOrderInputControl.

ecgan committed:
Rename MinimumOrderInputLabel to MinimumOrderInputLabelText.

ecgan committed:
Make edit work in EditMinimumOrderModal.

ecgan committed:
Make delete work in EditMinimumOrderModal.

ecgan committed:
Fix NaN threshold by checking free_shipping_threshold before calling Number().

ecgan committed:
Set threshold to undefined to prevent 0.

ecgan committed:
Code restructure with FreeShippingCards.

This is so that when users remove all the shipping rates, the OfferFreeShippingCard and MinimumOrderCard will not be rendered.

ecgan committed:
Add code for renderGroups in MinimumOrderCard.

ecgan committed:
Code refactor for renderGroups in MinimumOrderCard.

ecgan committed:
Allow add minimum order.

ecgan committed:
Fix filtering shipping rates in MinimumOrderCard.

ecgan committed:
Fix edit handler in MinimumOrderCard.

ecgan committed:
Better card title layout.

ecgan committed:
Better card title layout.

(cherry picked from commit 67f437283ea81160fd87df366e2fc1fbfab6c3cc)

ecgan committed:
Make WCDL SelectControl label look like other input label.

ecgan committed:
Make WCDL SelectControl label look like other input label.

(cherry picked from commit 2f4b15418b05d5d8e1020c5bc18fc04586952ac0)

ecgan committed:
Simplify SelectControl label in EditRateModal.

(cherry picked from commit 91d9310684a76673457dd0146b767c9938e457fe)

ecgan committed:
Simplify SelectControl label in EditRateModal.

ecgan committed:
Simplify layout code for add and edit rate modals.

ecgan committed:
Move AddRateModal and EditRateModal files.

ecgan committed:
Move files into minimum-order-card folder.

ecgan committed:
Replace AddRateButton with AppButtonModalTrigger.

ecgan committed:
Replace EditRateButton with AppButtonModalTrigger.

ecgan committed:
Refactor groupShippingRatesByFreeShippingThreshold to separate file.

ecgan committed:
Remove unneeded audienceCountries check and AppSpinner render.

ecgan committed:
Rename CountriesPriceInput to ShippingRateInputControl.

ecgan committed:
Make ShippingCountriesForm render card on its own and remove unnecessary ShippingRateSetup intermediate component.

ecgan committed:
Rename ShippingCountriesForm to EstimatedShippingRatesCard, and move files.

ecgan committed:
Fix export error for MinimumOrderCard.

ecgan committed:
FIx notice should be success, not error.

ecgan committed:
Remove obsolete code comments.

ecgan committed:
Merge branch 'develop' into feature/refactor-shipping-rates-api

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-api-ui-changes

ecgan committed:
Change export default for FlatShippingRatesInputCards.

ecgan committed:
Fix layout for the add rate button.

ecgan committed:
Move useShippingRatesWithSavedSuggestions files, to be used in SetupFreeListings.

ecgan committed:
Use FlatShippingRatesInputCards in Setup MC, replacing ShippingRateSetup.

mikkamp committed:
Helper function for returning last inserted ID

mikkamp committed:
Return updated or inserted rate in response

ecgan committed:
Merge branch 'tweak/return-rate-after-upsert' into feature/shipping-rates-api-ui-changes

ecgan committed:
Change upsertShippingRates based on new API changes.

New API changes is from https://github.com/woocommerce/google-listings-and-ads/pull/1249. Upsert now returns shipping rates objects with id.

ecgan committed:
Add saveShippingRates and automatically save shipping rates in Setup MC.

ecgan committed:
Use useDebouncedCallbackEffect in useAutoSaveShippingRatesEffect.

ecgan committed:
Use useIsEqualRefValue in useDebouncedCallbackEffect.

This is so that the effect is not triggered when value is still equal.

ecgan committed:
Change first render useRef in useDebouncedCallbackEffect.

This is because if we were to assign value to current, and if value is falsey value, it may not work as expected.

We change ref to boolean for better clarity.

ecgan committed:
Auto save settings when settings are changed, not when shipping rates are changed.

ecgan committed:
Better implementation for saveShippingRates.

The old implementation was deleting all the old shipping rates and upserting new ones, causing shipping rate ids to keep on incrementing.

In the new implementation, we selectively delete the old ones, and upsert the new ones. This prevents the shipping rate ids from keep on incrementing.

ecgan committed:
Delete unused ShippingRateSetup and child components.

ecgan committed:
Fix groupShippingRatesByPriceCurrency test.

ecgan committed:
Remove unneeded rates in groupShippingRatesByPriceCurrency.

ecgan committed:
Fix useShippingRatesSuggestions test based on shipping rates suggestions API changes.

The API changes are introduced in https://github.com/woocommerce/google-listings-and-ads/pull/1202.

ecgan committed:
Fix failed shipping rates reducer tests.

ecgan committed:
Fix CSS linting.

ecgan committed:
Update shipping rates schema in reducer test.

ecgan committed:
Fix deleteShippingRates in useAutoClearShippingEffect.

ecgan committed:
Check for ids in deleteShippingRates.

This is to be more defensive and prevent API error response when sending empty ids array.

mikkamp committed:
Rename query variable

The query can be used for update or insert, so calling it
`$update_query` is confusing.

ecgan committed:
Save form values when users click on Continue button.

mikkamp committed:
Store a copy of the last inserted ID

This is an extra precaution in case the last inserted ID is not fetched
right away after the insert.

ecgan committed:
Code refactor with isSameShippingRate.

ecgan committed:
Merge branch 'develop' into feature/refactor-shipping-rates-api

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-api-ui-changes

mikkamp committed:
Merge pull request #1249 from woocommerce/tweak/return-rate-after-upsert

Return the updated or inserted rate in the response

ecgan committed:
Fix shipping rates not saved correctly in Edit Free Listings.

When users delete shipping rates, click on save changes button, and reload the page, the shipping rates would show up again and did not get deleted. This is because we were calling upsertShippingRates, and there is no delete.

ecgan committed:
Make ShippingRateSection shared component and use in Setup MC and Edit Free Listings.

ecgan committed:
Move FlatShippingRatesInputCards into ShippingRateSection component folder.

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-api-ui-changes

ecgan committed:
Standardize on saveShippingRates, remove upsertShippingRates and deleteShippingRates.

ecgan committed:
Use RECEIVE_SHIPPING_RATES, as SAVE_SHIPPING_RATES is not necessary.

ecgan committed:
Code refactor with SHIPPING_RATE_METHOD.FLAT_RATE.

ecgan committed:
Code refactor with SHIPPING_RATE_METHOD.FLAT_RATE.

(cherry picked from commit 78dfd29fb7c9bf5dfe2f250ea0c33e998c4fa218)

ecgan committed:
Defensive check on shipping rate flat_rate method.

ecgan committed:
Defensive check on shipping rate flat_rate method.

(cherry picked from commit 13f3255f60400417a2389ece8b110dcf331a9b50)

ecgan committed:
Change groupShippingRatesByPriceCurrency to also group by shipping rate method.

ecgan committed:
Change groupShippingRatesByPriceCurrency to also group by shipping rate method.

(cherry picked from commit 05968b4a592dab68f341ba74270c3d2bf9fc9f45)

ecgan committed:
Rename groupShippingRatesByPriceCurrency to groupShippingRatesByMethodPriceCurrency.

ecgan committed:
Fix error with missing shipping rate method.

The error happens when we go to EditRateModal and add a country and save the changes.

ecgan committed:
Fix error with missing shipping rate method.

The error happens when we go to EditRateModal and add a country and save the changes.

(cherry picked from commit 34e7588e21a5b83313e8cca7ac23d094816871b2)

ecgan committed:
Standardize the schema in countriesPriceArray.

ecgan committed:
Change and rename groupShippingRatesByFreeShippingThreshold to also group by method.

ecgan committed:
Fix checkErrors shippingRates value should come from form value.

ecgan committed:
Change shipping rates code in wp-data.

Upsert and delete shipping rates are not really needed, we only need saveShippingRates.

ecgan committed:
Change upsertShippingRates to saveShippingRates in EditFreeCampaign.

ecgan committed:
Remove obsolete code comments.

ecgan committed:
Fix notice should be success, not error, when saving is successful.

ecgan committed:
Merge branch 'feature/shipping-rates-ui/select-control-label' into feature/shipping-rates-ui/edit-free-listings

Conflicts:
	js/src/components/flat-shipping-rates-input-cards/estimated-shipping-rates-card/edit-rate-modal.js
	js/src/components/free-listings/configure-product-listings/shipping-rate/shipping-rate-setup/index.js
	js/src/data/actions.js
	js/src/data/reducer.js
	js/src/edit-free-campaign/index.js

ecgan committed:
Revert files in Setup MC, to have a better clean PR.

ecgan committed:
Make ShippingRateSection shared component and move FlatShippingRatesInputCards into it.

ecgan committed:
Rename ShippingRateSection countries prop to audienceCountries.

This is to make it consistent with child components.

ecgan committed:
CSS lint fix.

ecgan committed:
Better component import.

ecgan committed:
Rename groupShippingRatesByPriceCurrency to groupShippingRatesByMethodPriceCurrency.

ecgan committed:
Merge branch 'feature/shipping-rates-ui/edit-free-listings' into feature/shipping-rates-ui/setup-mc

Conflicts:
	js/src/components/flat-shipping-rates-input-cards/minimum-order-card/groupShippingRatesByFreeShippingThreshold.js
	js/src/components/shipping-rate-section/minimum-order-card/groupShippingRatesByFreeShippingThreshold.js
	js/src/components/shipping-rate-section/minimum-order-card/groupShippingRatesByMethodFreeShippingThreshold.js
	js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/add-rate-button/add-rate-modal/index.js
	js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/countries-price-input/edit-rate-button/edit-rate-modal/index.js
	js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/index.js
	js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/index.scss
	js/src/setup-mc/setup-stepper/setup-free-listings/useSaveSuggestions.js

ecgan committed:
Use Section.Card.Title in shipping time section in Setup MC.

ecgan committed:
Update JSDoc for shipping rate.

ecgan committed:
Convert the upsertedShippingRates' rate into number.

This is to make saveShippingRates consistent with fetchShippingRates.

ecgan committed:
Code refactor with isInShippingRates.

ecgan committed:
Refactor saveShippingRates into smaller functions.

ecgan committed:
Update code comments.

ecgan committed:
Remove unneeded optional chaining.

ecgan committed:
Remove unused ShippingRateSetup formProps in Setup MC.

ecgan committed:
Change card title margin bottom to 16px.

ecgan committed:
Remove unneeded div.

ecgan committed:
Use AppCountrySelect label props in AddRateModal in Edit Free Listings.

ecgan committed:
Merge branch 'feature/shipping-rates-ui/select-control-label' into feature/shipping-rates-ui/edit-free-listings

ecgan committed:
Remove options from isInShippingRates test.

`options` is supposed to be an object, but when there are no options, the API returned it as an array, which can be confusing. Removing the options array here for now.

ecgan committed:
Merge pull request #1262 from woocommerce/feature/shipping-rates-ui/wp-data

ecgan committed:
Merge pull request #1263 from woocommerce/feature/shipping-rates-ui/remove-old-settings

ecgan committed:
Merge pull request #1264 from woocommerce/feature/shipping-rates-ui/card-title-layout

ecgan committed:
Merge pull request #1265 from woocommerce/feature/shipping-rates-ui/select-control-label

Commit:
Cast empty arrays to objects if property type is object.

This is to avoid including an empty `array` in the API response in case the property type is `object` but it has no properties (i.e. no array key-values).

An example of this is in the `GET mc/shipping/rates` API where the `options` property of each shipping rate could be returned as an empty array.

ecgan committed:
Remove unneeded optional chaining in FreeShippingCards.

ecgan committed:
Fix missed renaming for groupShippingRatesByMethodPriceCurrency.

ecgan committed:
Remove unneeded Number() conversion in groupShippingRatesByMethodPriceCurrency.

ecgan committed:
Code refactor with isNonZeroFlatShippingRate.

ecgan committed:
Simplify groupShippingRatesByFreeShippingThreshold.

We don't need optional chaining, and free_shipping_threshold is always number from the API.

ecgan committed:
Better validation for minimum order input.

Check for threshold must be greater than 0.

ecgan committed:
Merge branch 'feature/shipping-rates-ui/main' into feature/shipping-rates-ui/edit-free-listings

Commit:
Add unit tests

nima-karimi committed:
Merge pull request #1284 from woocommerce/feature/return-empty-object-api

Cast empty arrays to objects in APIs if property type is object.

ecgan committed:
Rename isNonZeroFlatShippingRate to isNonFreeFlatShippingRate.

ecgan committed:
Make AddMinimumOrderModal display options using countries with no threshold value.

This is to match with AddRateModal behavior.

ecgan committed:
Better HTML output with sprintf in MinimumOrderInputLabelText.

ecgan committed:
Change options from [] to {}.

This is in accordance to https://github.com/woocommerce/google-listings-and-ads/pull/1284.

ecgan committed:
Merge branch 'feature/shipping-rates-ui/edit-free-listings' into feature/shipping-rates-ui/setup-mc

Conflicts:
	js/src/components/shipping-rate-section/minimum-order-card/groupShippingRatesByMethodFreeShippingThreshold.js
	js/src/components/shipping-rate-section/minimum-order-card/minimum-order-card.js
	js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate-section.js

ecgan committed:
SImplify code in groupShippingRatesByMethodFreeShippingThreshold.

ecgan committed:
Change options from [] to {}.

This is as per the changes in https://github.com/woocommerce/google-listings-and-ads/pull/1284.

ecgan committed:
Merge pull request #1266 from woocommerce/feature/shipping-rates-ui/edit-free-listings

ecgan committed:
Fix code comment in MinimumOrderCard.

ecgan committed:
Remove "all countries" from label and replace with ShippingRateInputControlLabelText.

`totalCountyCount` has been removed because it is not needed anymore.

ecgan committed:
Code refactor with CountryNamesPlusMore.

ecgan committed:
Fix import using project root path.

ecgan committed:
Use ShippingTimeInputControlLabelText in shipping times section.

totalCountyCount has been removed because it is no longer in use.

ecgan committed:
Remove unused CountryNames component.

Commit:
Fix the onboarding step returned as shipping after saving rates

The API to return the current onboarding step did not recognize the newly added `automatic` mode for syncing shipping settings and also the query to count the countries that have a saved shipping rate was not working as expected.

ecgan committed:
Fix thresholdGroups containing groups with undefined threshold.

ecgan committed:
Merge pull request #1267 from woocommerce/feature/shipping-rates-ui/setup-mc

ecgan committed:
Add offer_free_shipping form key to Edit Free Listings.

ecgan committed:
Put shipping rate and offer free shipping change handlers in FlatShippingRatesInputCards.

Some of the logics are moved from FreeShippingCards into FlatShippingRatesInputCards. We don't need the useState in FreeShippingCards, since the state is now kept in form using offer_free_shipping. FreeShippingCards is removed as it is not needed anymore.

ecgan committed:
Code refactor with getOfferFreeShippingInitialValue.

ecgan committed:
Add offer_free_shipping form key in Setup MC.

ecgan committed:
Set options based on price.

When shipping rate is changed to free shipping, we set options to empty object to clear the free shipping threshold.

ecgan committed:
Simplify code.

Commit:
Serialize options using JSON instead of php serializer

Commit:
Move existing free shipping threshold to rates table

Commit:
Check method is flat_rate

Commit:
Check `offers_free_shipping` is not false

Commit:
Check the saved shipping times match target audience

Instead of just checking whether the number of saved countries, we also check the name of the countries to make sure that all target audience countries have a shipping time saved in DB.

nima-karimi committed:
Merge pull request #1298 from woocommerce/feature/invalid-onboarding-step-shipping

Fix the current onboarding step being returned as shipping even after saving rates

Commit:
Merge branch 'develop' into feature/refactor-shipping-rates-api

# Conflicts:
#	src/Internal/DependencyManagement/CoreServiceProvider.php
#	src/MerchantCenter/MerchantCenterService.php

Commit:
Fix merge conflict

eason9487 committed:
Compare the key fields to check if there are unsaved shipping rates in free listings edit.

ecgan committed:
Merge pull request #1303 from woocommerce/feature/shipping-rates-ui/clear-free-shipping-threshold

ecgan committed:
Delete unused more.js file.

ecgan committed:
Rename MinimumOrderInputLabelText to MinimumOrderInputControlLabelText.

This is to make it consistent with other InputControlLabelText.

ecgan committed:
Merge pull request #1297 from woocommerce/feature/shipping-rates-ui/remove-all-countries-label

ecgan committed:
Rename functions to indicate "get change handler".

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-ui/main

eason9487 committed:
Fix the naming typos of `hasUnsavedShippingRates` function, file and path.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1307#discussion_r822944729

eason9487 committed:
Merge pull request #1307 from woocommerce/fix/1290-prompt-unsaved-shipping-rates

Shipping rates UI changes: Adjust the logic of prompting for unsaved shipping rates

Commit:
Decode options JSON as array

Commit:
Always remove previous free shipping options

Commit:
Rename serialized options variable

Commit:
Move phpcs disable comments

nima-karimi committed:
Merge pull request #1305 from woocommerce/feature/migrate-free-shipping-option

Migrate free shipping threshold settings to the shipping rates table

ecgan committed:
Add validation for offer_free_shipping field.

ecgan committed:
Code refactor to use values.shipping_country_rates in checkErrors.

We don't need shippingRates as function parameter anymore, since shippingRates is now in the values object as values.shipping_country_rates.

Eventually we should also do the same for shippingTimes.

ecgan committed:
Merge pull request #1302 from woocommerce/feature/shipping-rates-ui/offer-free-shipping

Commit:
Use constants from interface instead of options class

Commit:
Remove unused class

ecgan committed:
Fix checkErrors tests.

ecgan committed:
Add tests for offer free shipping in checkErrors.

ecgan committed:
Move offrer free shipping code into its own section in checkErrors.

ecgan committed:
Change saveShippingRates from wp.data action to custom hook.

Also adds upsertShippingRates and deleteShippingRates in wp.data used by saveShippingRates.

ecgan committed:
Replace usage of wp.data saveShippingRates action with the custom hook version.

ecgan committed:
Add error handling in useAutoClearShippingEffect.

This is because saveShippingRates does not have error handling in it by design.

ecgan committed:
Code refactor with getDeleteIds.

ecgan committed:
Change hasUnsavedShippingRates to getUnsavedShippingRates.

This is so that we can use getUnsavedShippingRates later for saveShippingRates operation.

ecgan committed:
Fix getUnsavedShippingRates returning boolean wrongly.

ecgan committed:
Make upsertShippingRates to be called for unsaved shipping rates only.

ecgan committed:
Fix navigate away prompt showing up when we change offer_free_shipping value back and forth.

newSettings should not contain offer_free_shipping, so we use a new function getSettings here to extract the settings data from the form values object, leaving offer_free_shipping out.

ecgan committed:
Change offer_free_shipping error message text.

ecgan committed:
Fix validation for offer free shipping should only be done for flat shipping settings.

ecgan committed:
Merge pull request #1309 from woocommerce/feature/shipping-rates-ui/validation-offer-free-shipping

ecgan committed:
Revert changes to hasUnsavedShippingRates.

This reverts commit 2da3806088c111c7d4fda16788dde2b447580551, 7eab10a78ff8c6202762fe067a9e971aca9d22d8 and 53058fd8b5e35d7cbd7632733018128c6566ed6f.

ecgan committed:
Add tests for hasUnsavedShippingRates.

ecgan committed:
Add code comment for getSettings.

ecgan committed:
Simplify handleChange, don't need to destructure values.

ecgan committed:
Add code comment for offer_free_shipping.

This is to indicate offer_free_shipping does not participate in API calls.

ecgan committed:
Use lodash pick in getSettings.

This is to make the code more succinct and less verbose.

ecgan committed:
Add code comment.

ecgan committed:
Change switch to if-else.

We don't want to use the default case because it would match the offer_free_shipping changes, which isn't right. Instead, we check with settingsFieldNames.includes.

We don't need to do anything for offer_free_shipping in handleChange.

ecgan committed:
Edit code comments.

ecgan committed:
Code refactor with getDifferentShippingRates.

ecgan committed:
Edit code comment for getDifferentShippingRates.

ecgan committed:
Use getDifferentShippingRates in saveShippingRates.

ecgan committed:
Code refactor for getDeleteIds.

ecgan committed:
setSaving to false only when there is an error.

ecgan committed:
Error handling in useAutoSaveShippingRatesEffect.

ecgan committed:
Add reducer tests for upsert and delete shipping rates.

ecgan committed:
Add JSDoc for upsertShippingRates and deleteShippingRates.

ecgan committed:
Code refactor with getDeletedShippingRates.

ecgan committed:
Fix missing async await.

ecgan committed:
Update code comment.

This is based on code review in https://github.com/woocommerce/google-listings-and-ads/pull/1317#pullrequestreview-911072881.

ecgan committed:
Merge pull request #1317 from woocommerce/feature/shipping-rates-ui/fix-navigate-away-offer-free-shipping

ecgan committed:
Add and change tests for hasUnsavedShippingRates.

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-ui/main

ecgan committed:
Fix test description for hasUnsavedShippingRates.

Co-authored-by: Eason <eason.su.tw@gmail.com>

ecgan committed:
Remove unneeded useCallback.

We don't need useCallback because that is already handled by useDebouncedCallbackEffect.

ecgan committed:
Merge pull request #1313 from woocommerce/feature/shipping-rates-ui/saveShippingRates-custom-hook

ecgan committed:
Merge pull request #1322 from woocommerce/feature/shipping-rates-ui/main

Commit:
Merge branch 'develop' into feature/refactor-shipping-rates-api

# Conflicts:
#	src/API/Site/Controllers/CountryCodeTrait.php

Commit:
Fix merge conflict

ecgan committed:
Filter out shipping rates suggestions that are not flat_rate method.

ecgan committed:
Merge pull request #1367 from woocommerce/feature/remove-free-shipping-suggestions
